File: src/core/simulation-engine.js
/**
 * Central Simulation Engine
 * Master clock and state coordinator uniting Workload, Smart Cache, DB, Latencies, Costs, Scaling, and Telemetry.
 */

import { WORKLOAD_TYPES, TRAFFIC_SCENARIOS, CACHE_STRATEGIES, DEFAULT_CONFIG } from './types.js';
import { globalEventBus } from './event-bus.js';
import { ItemCatalog } from '../workload/item-catalog.js';
import { WorkloadGenerator } from '../workload/generator.js';
import { SmartCache } from '../cache/smart-cache.js';
import { LRUCache } from '../cache/lru-cache.js';
import { LFUCache } from '../cache/lfu-cache.js';
import { GDSCache } from '../cache/gds-cache.js';
import { DatabaseSimulator } from '../backend/database-sim.js';
import { BackendService } from '../backend/backend-service.js';
import { CostModel } from '../cost/cost-model.js';
import { ScalingAdvisor } from '../scaling/scaling-advisor.js';
import { MetricsEngine } from '../metrics/metrics-engine.js';
import { HistoryRecorder } from '../timemachine/history-recorder.js';

export class SimulationEngine {
  constructor() {
    this.catalog = new ItemCatalog();
    this.workloadGenerator = new WorkloadGenerator(this.catalog);
    
    this.activeStrategy = CACHE_STRATEGIES.SMART;
    this.cacheCapacityBytes = DEFAULT_CONFIG.cacheCapacityBytes;
    
    // Cache Strategy Instances
    this.smartCache = new SmartCache(this.cacheCapacityBytes, WORKLOAD_TYPES.READ_HEAVY_API);
    this.lruCache = new LRUCache(this.cacheCapacityBytes);
    this.lfuCache = new LFUCache(this.cacheCapacityBytes);
    this.gdsCache = new GDSCache(this.cacheCapacityBytes);

    // Backend & Infrastructure Simulators
    this.dbSimulator = new DatabaseSimulator();
    this.backendService = new BackendService();
    this.costModel = new CostModel();
    this.scalingAdvisor = new ScalingAdvisor();
    this.metricsEngine = new MetricsEngine();
    this.historyRecorder = new HistoryRecorder();

    // Simulation Execution State
    this.isRunning = true;
    this.tickIntervalMs = DEFAULT_CONFIG.tickIntervalMs;
    this.timerId = null;
    this.currentSimTime = 0;
    this.lastTickTimestamp = Date.now();

    // Warm-up initial items so the dashboard starts with realistic cached data
    this.seedWarmCache();
  }

  seedWarmCache() {
    const items = this.catalog.getCatalog(this.workloadGenerator.workloadType);
    for (let i = 0; i < Math.min(35, items.length); i++) {
      const item = items[i];
      this.smartCache.put(item, 0);
      this.lruCache.put(item, 0);
      this.lfuCache.put(item, 0);
      this.gdsCache.put(item, 0);
    }
  }

  start() {
    if (this.isRunning && this.timerId) return;
    this.isRunning = true;
    this.lastTickTimestamp = Date.now();
    this.timerId = setInterval(() => this.tick(), this.tickIntervalMs);
    globalEventBus.emit('simulation_status', { isRunning: true });
  }

  pause() {
    this.isRunning = false;
    if (this.timerId) {
      clearInterval(this.timerId);
      this.timerId = null;
    }
    globalEventBus.emit('simulation_status', { isRunning: false });
  }

  step() {
    if (!this.isRunning) {
      this.tick();
    }
  }

  reset() {
    this.pause();
    this.currentSimTime = 0;
    this.smartCache.clear();
    this.lruCache.clear();
    this.lfuCache.clear();
    this.gdsCache.clear();
    this.dbSimulator.reset();
    this.backendService.reset();
    this.metricsEngine.reset();
    this.historyRecorder.reset();
    this.seedWarmCache();
    this.tick();
    globalEventBus.emit('simulation_reset', {});
  }

  setWorkloadType(type) {
    this.workloadGenerator.setWorkloadType(type);
    this.smartCache.setWorkloadType(type);
    globalEventBus.emit('decision_feed_event', {
      type: 'WORKLOAD_SWITCH',
      title: `Switched Workload to ${type === WORKLOAD_TYPES.READ_HEAVY_API ? 'Read-Heavy API' : 'Compute-Heavy Recommendation'}`,
      description: `Adapted multi-factor weights and recomputation latency models.`,
      severity: 'info',
      timestamp: this.currentSimTime
    });
    this.tick();
  }

  setScenario(scenario, options = {}) {
    this.workloadGenerator.setScenario(scenario, options);
    
    let scenarioDesc = `Traffic pattern switched to ${scenario}`;
    if (scenario === TRAFFIC_SCENARIOS.POPULARITY_SPIKE) {
      scenarioDesc = `Popularity surge triggered on ${this.workloadGenerator.spikeTargetItemId} (+4500% request share).`;
    } else if (scenario === TRAFFIC_SCENARIOS.COLD_START) {
      this.smartCache.clear();
      this.lruCache.clear();
      this.lfuCache.clear();
      this.gdsCache.clear();
      scenarioDesc = `Cache wiped. Cold start initiated under full incoming traffic.`;
    } else if (scenario === TRAFFIC_SCENARIOS.CACHE_POLLUTION) {
      scenarioDesc = `Unique key deluge injected. Testing Cache Pollution Defense.`;
    } else if (scenario === TRAFFIC_SCENARIOS.TRAFFIC_BURST) {
      scenarioDesc = `Traffic volume surge: 3.8x baseline request burst.`;
    }

    globalEventBus.emit('decision_feed_event', {
      type: 'SCENARIO_CHANGE',
      title: `Scenario: ${scenario.replace('_', ' ')}`,
      description: scenarioDesc,
      severity: scenario === TRAFFIC_SCENARIOS.CACHE_POLLUTION ? 'warning' : 'success',
      timestamp: this.currentSimTime
    });

    this.tick();
  }

  setBaseRps(rps) {
    this.workloadGenerator.setBaseRps(rps);
    this.tick();
  }

  setSpeedMultiplier(speed) {
    this.workloadGenerator.setSpeedMultiplier(speed);
  }

  setCapacity(newCapacityBytes) {
    this.cacheCapacityBytes = newCapacityBytes;
    this.smartCache.setCapacity(newCapacityBytes);
    this.lruCache.setCapacity(newCapacityBytes);
    this.lfuCache.setCapacity(newCapacityBytes);
    this.gdsCache.setCapacity(newCapacityBytes);

    globalEventBus.emit('decision_feed_event', {
      type: 'CAPACITY_CHANGE',
      title: `Cache Capacity Resized to ${(newCapacityBytes / (1024 * 1024 * 1024)).toFixed(1)} GB`,
      description: `Eviction thresholds and memory allocation recalculated.`,
      severity: 'info',
      timestamp: this.currentSimTime
    });

    this.tick();
  }

  /**
   * Main Simulation Step
   */
  tick() {
    const deltaMs = this.tickIntervalMs;
    
    // 1. Generate Request Batch
    const batch = this.workloadGenerator.tick(deltaMs);
    this.currentSimTime = batch.simTimeSeconds;
    const { requests, effectiveRps, workloadType } = batch;

    // 2. Process requests through Smart Cache and shadow stores
    const batchResults = [];
    let tickHits = 0;
    let tickMisses = 0;

    for (const req of requests) {
      // Primary Smart Cache evaluation
      let getRes = this.smartCache.get(req.id, this.currentSimTime);
      if (!getRes.hit) {
        this.smartCache.put(req, this.currentSimTime);
        tickMisses++;
      } else {
        tickHits++;
      }
      batchResults.push({ hit: getRes.hit, item: req });

      // Shadow runs for LRU, LFU, GDS
      let lruRes = this.lruCache.get(req.id, this.currentSimTime);
      if (!lruRes.hit) this.lruCache.put(req, this.currentSimTime);

      let lfuRes = this.lfuCache.get(req.id, this.currentSimTime);
      if (!lfuRes.hit) this.lfuCache.put(req, this.currentSimTime);

      let gdsRes = this.gdsCache.get(req.id, this.currentSimTime);
      if (!gdsRes.hit) this.gdsCache.put(req, this.currentSimTime);
    }

    // 3. Database Simulation
    const dbTelemetry = this.dbSimulator.processMisses(tickMisses, deltaMs, workloadType);

    // 4. Backend Service Simulation & Latency generation
    const backendResult = this.backendService.processRequestBatch(batchResults, dbTelemetry, workloadType);
    this.metricsEngine.addLatencySamples(backendResult.latencies);

    // 5. Periodic Smart Cache Maintenance
    this.smartCache.periodicMaintenance(this.currentSimTime, requests.length);

    // 6. Compute Telemetry & Percentiles
    const percentiles = this.metricsEngine.computePercentiles();
    const hitRate = this.smartCache.getHitRate();
    const missRate = 1 - hitRate;

    const costTelemetry = {
      cacheCapacityBytes: this.cacheCapacityBytes,
      usedBytes: this.smartCache.usedBytes,
      hitsPerSecond: effectiveRps * hitRate,
      missesPerSecond: effectiveRps * missRate,
      backendLoadPercent: backendResult.backendLoadPercent,
      dbQueriesPerSecond: effectiveRps * missRate,
      totalRequestsWindow: this.smartCache.hits + this.smartCache.misses,
      totalHitsWindow: this.smartCache.hits,
      totalMissesWindow: this.smartCache.misses
    };

    const costResult = this.costModel.computeCost(costTelemetry);
    const scalingResult = this.scalingAdvisor.evaluateScaling(
      this.cacheCapacityBytes,
      {
        hitsPerSecond: costTelemetry.hitsPerSecond,
        missesPerSecond: costTelemetry.missesPerSecond,
        hitRate,
        backendLoadPercent: backendResult.backendLoadPercent,
        memoryUsagePercent: this.smartCache.getUsagePercent()
      }
    );

    const pollutionStatus = this.smartCache.pollutionGuard.getStatus();

    const snapshot = {
      simTime: Math.round(this.currentSimTime),
      trafficRps: effectiveRps,
      hitRatePercent: Number((hitRate * 100).toFixed(1)),
      missRatePercent: Number((missRate * 100).toFixed(1)),
      p50: percentiles.p50,
      p95: percentiles.p95,
      p99: percentiles.p99,
      avgLatency: percentiles.avg,
      costPerHour: costResult.totalCostPerHour,
      costSavingsPerHour: costResult.costSavingsPerHour,
      savingsPercentage: costResult.savingsPercentage,
      uncachedCostPerHour: costResult.uncachedTotalCostPerHour,
      cacheCostPerHour: costResult.cacheCostPerHour,
      computeCostPerHour: costResult.computeCostPerHour,
      dbCostPerHour: costResult.dbCostPerHour,
      backendLoadPercent: backendResult.backendLoadPercent,
      activeThreads: backendResult.activeThreads,
      dbCpuPercent: dbTelemetry.cpuUtilizationPercent,
      dbLatencyMs: dbTelemetry.currentLatencyMs,
      dbConnections: dbTelemetry.activeConnections,
      dbQueriesPerSecond: dbTelemetry.queriesPerSecond,
      memoryUsedMB: Number((this.smartCache.usedBytes / (1024 * 1024)).toFixed(1)),
      memoryCapacityMB: Number((this.cacheCapacityBytes / (1024 * 1024)).toFixed(1)),
      memoryUsagePercent: Number(this.smartCache.getUsagePercent().toFixed(1)),
      evictions: this.smartCache.evictions,
      refreshes: this.smartCache.refreshes,
      activeItemsCount: this.smartCache.entries.size,
      workloadType,
      scenario: batch.scenario,
      pollutionRisk: pollutionStatus.riskLevel,
      uniqueKeyRatePercent: pollutionStatus.uniqueKeyRatePercent,
      usefulOccupancyPercent: pollutionStatus.usefulOccupancyPercent,
      scalingDecision: scalingResult
    };

    // Record in metrics history and time machine
    this.metricsEngine.recordSnapshot(snapshot);
    this.historyRecorder.recordState(this.currentSimTime, snapshot);

    // Broadcast update to UI
    globalEventBus.emit('simulation_tick', {
      snapshot,
      smartCache: this.smartCache,
      lruCache: this.lruCache,
      lfuCache: this.lfuCache,
      gdsCache: this.gdsCache,
      scalingResult,
      pollutionStatus
    });
  }
}

File: backend/core/simulation_engine.py
"""
Central Simulation Engine (Python Core)
Master coordinator uniting Workload Generator, Smart Cache, Baseline Caches, DB, Latencies, Traffic Monitor, Predictive Pre-Warmer, Cost Model, Scaling Controller, and Decision Logger.
"""

from collections import deque
import time
from typing import Dict, Any, List, Optional
from backend.core.types import (
    WorkloadType, TrafficScenario, CacheStrategy, DecisionType,
    SystemConfig, ScoringWeights
)
from backend.workload.catalog import ItemCatalog
from backend.workload.generator import WorkloadGenerator
from backend.cache.smart_cache import SmartCache
from backend.cache.lru_cache import LRUCache
from backend.cache.lfu_cache import LFUCache
from backend.cache.gds_cache import GDSCache
from backend.services.database_sim import DatabaseSimulator, BackendService
from backend.services.traffic_monitor import TrafficMonitor
from backend.services.predictive_warmer import PredictiveWarmer
from backend.services.cost_model import CostModel
from backend.services.scaling_controller import ScalingController
from backend.services.decision_logger import DecisionLogger


class SimulationEngine:
    def __init__(self):
        self.catalog = ItemCatalog()
        self.workload_generator = WorkloadGenerator(self.catalog)
        self.cache_capacity_bytes = SystemConfig.cacheCapacityBytes

        # Cache instances
        self.smart_cache = SmartCache(self.cache_capacity_bytes, WorkloadType.READ_HEAVY_API)
        self.lru_cache = LRUCache(self.cache_capacity_bytes)
        self.lfu_cache = LFUCache(self.cache_capacity_bytes)
        self.gds_cache = GDSCache(self.cache_capacity_bytes)

        # Services
        self.db_simulator = DatabaseSimulator()
        self.backend_service = BackendService()
        self.traffic_monitor = TrafficMonitor()
        self.predictive_warmer = PredictiveWarmer()
        self.cost_model = CostModel()
        self.scaling_controller = ScalingController()
        self.decision_logger = DecisionLogger()

        # Telemetry & history
        self.is_running = True
        self.current_sim_time = 0.0
        self.latency_reservoir: deque = deque(maxlen=250)
        self.rolling_history: deque = deque(maxlen=40)
        self.last_scaling_action_time = 0.0

        # Seed initial warm cache entries
        self.seed_warm_cache()

    def seed_warm_cache(self):
        items = self.catalog.get_catalog(self.workload_generator.workload_type)
        for i in range(min(35, len(items))):
            item = items[i]
            self.smart_cache.put(item, 0.0)
            self.lru_cache.put(item, 0.0)
            self.lfu_cache.put(item, 0.0)
            self.gds_cache.put(item, 0.0)

    def set_workload_type(self, w_type: WorkloadType):
        self.workload_generator.set_workload_type(w_type)
        self.smart_cache.set_workload_type(w_type)
        self.decision_logger.log_decision(
            event_type="WORKLOAD_SWITCH",
            title=f"Switched Workload to {'Read-Heavy API' if w_type == WorkloadType.READ_HEAVY_API else 'Compute-Heavy Recommendation'}",
            description="Adapted multi-factor weights, recompute costs, and dynamic latency models.",
            severity="info",
            timestamp=self.current_sim_time
        )

    def set_scenario(self, scenario: TrafficScenario, options: Optional[Dict[str, Any]] = None):
        self.workload_generator.set_scenario(scenario, options)
        
        desc = f"Traffic pattern switched to {scenario.value}"
        if scenario == TrafficScenario.POPULARITY_SPIKE:
            desc = f"Popularity surge triggered on {self.workload_generator.spike_target_item_id} (+4500% request share)."
        elif scenario == TrafficScenario.COLD_START:
            self.smart_cache.clear()
            self.lru_cache.clear()
            self.lfu_cache.clear()
            self.gds_cache.clear()
            desc = "Cache wiped. Cold start initiated under full incoming traffic."
        elif scenario == TrafficScenario.CACHE_POLLUTION:
            desc = "Unique key deluge injected. Testing Cache Pollution Defense."
        elif scenario == TrafficScenario.TRAFFIC_BURST:
            desc = "Transient traffic volume surge: 3.8x baseline request burst."

        self.decision_logger.log_decision(
            event_type="SCENARIO_CHANGE",
            title=f"Scenario: {scenario.value.replace('_', ' ')}",
            description=desc,
            severity="warning" if scenario == TrafficScenario.CACHE_POLLUTION else "success",
            timestamp=self.current_sim_time
        )

    def set_base_rps(self, rps: int):
        self.workload_generator.set_base_rps(rps)

    def set_speed_multiplier(self, speed: float):
        self.workload_generator.set_speed_multiplier(speed)

    def set_capacity(self, new_capacity_bytes: int):
        self.cache_capacity_bytes = new_capacity_bytes
        self.smart_cache.set_capacity(new_capacity_bytes)
        self.lru_cache.set_capacity(new_capacity_bytes)
        self.lfu_cache.set_capacity(new_capacity_bytes)
        self.gds_cache.set_capacity(new_capacity_bytes)

        gb = new_capacity_bytes / (1024.0 * 1024.0 * 1024.0)
        self.decision_logger.log_decision(
            event_type="CAPACITY_CHANGE",
            title=f"Cache Capacity Resized to {gb:.1f} GB",
            description="Eviction thresholds and memory allocation recalculated.",
            severity="primary",
            timestamp=self.current_sim_time
        )

    def set_scoring_weights(self, weights_dict: Dict[str, float]):
        self.smart_cache.scorer.set_weights(weights_dict)
        self.decision_logger.log_decision(
            event_type="SCORING_WEIGHTS_UPDATED",
            title="Custom Scoring Weights Applied",
            description="Scoring formula recalculated for all resident objects.",
            severity="info",
            timestamp=self.current_sim_time
        )

    def set_auto_adapt_weights(self, enabled: bool):
        self.smart_cache.scorer.set_auto_adapt(enabled)

    def ingest_custom_data(self, items: List[Dict[str, Any]]):
        self.catalog.read_heavy_catalog = items
        self.catalog.compute_heavy_catalog = items
        self.workload_generator.recompute_weights()
        self.smart_cache.clear()
        self.lru_cache.clear()
        self.lfu_cache.clear()
        self.gds_cache.clear()
        self.seed_warm_cache()
        self.decision_logger.log_decision(
            event_type="USER_DATA_INGESTED",
            title="Custom User Data Ingested",
            description=f"Loaded {len(items)} custom items into active catalog & simulation pipeline.",
            severity="success",
            timestamp=self.current_sim_time
        )

    def reset(self):
        self.current_sim_time = 0.0
        self.smart_cache.clear()
        self.lru_cache.clear()
        self.lfu_cache.clear()
        self.gds_cache.clear()
        self.db_simulator.reset()
        self.backend_service.reset()
        self.decision_logger.clear()
        self.latency_reservoir.clear()
        self.rolling_history.clear()
        self.cache_capacity_bytes = SystemConfig.cacheCapacityBytes
        self.smart_cache.set_capacity(self.cache_capacity_bytes)
        self.seed_warm_cache()
        self.decision_logger.log_decision(
            event_type="SIMULATION_RESET",
            title="Simulation Environment Reset",
            description="All cache stores, latency reservoirs, and metrics cleared.",
            severity="info",
            timestamp=0.0
        )

    def tick(self, delta_ms: int = 250) -> Dict[str, Any]:
        # 1. Generate Request Batch
        batch = self.workload_generator.tick(delta_ms)
        self.current_sim_time = batch["simTimeSeconds"]
        requests = batch["requests"]
        effective_rps = batch["effectiveRps"]
        workload_type = batch["workloadType"]
        scenario = batch["scenario"]

        # 2. Process requests through Smart Cache and shadow stores
        batch_results = []
        tick_hits = 0
        tick_misses = 0

        for req in requests:
            req_id = req["id"]
            # Track rate for predictive pre-warming
            self.predictive_warmer.track_access(req_id)

            get_res = self.smart_cache.get(req_id, self.current_sim_time)
            if not get_res["hit"]:
                put_res = self.smart_cache.put(req, self.current_sim_time)
                tick_misses += 1
                if put_res.get("admitted") and put_res.get("entry"):
                    e = put_res["entry"]
                    if len(self.decision_logger.events) < 5 or tick_misses % 8 == 0:
                        self.decision_logger.log_cache_action(
                            action=e.decision,
                            item_id=req_id,
                            score=e.score,
                            reason=e.reason,
                            factors=e.factors,
                            timestamp=self.current_sim_time
                        )
            else:
                tick_hits += 1
                e = get_res["entry"]
                if e and e.decision == DecisionType.REFRESH.value:
                    self.decision_logger.log_cache_action(
                        action=DecisionType.REFRESH.value,
                        item_id=req_id,
                        score=e.score,
                        reason=e.reason,
                        factors=e.factors,
                        timestamp=self.current_sim_time
                    )

            batch_results.append({"hit": get_res["hit"], "item": req})

            # Shadow caches
            lru_res = self.lru_cache.get(req_id, self.current_sim_time)
            if not lru_res["hit"]:
                self.lru_cache.put(req, self.current_sim_time)

            lfu_res = self.lfu_cache.get(req_id, self.current_sim_time)
            if not lfu_res["hit"]:
                self.lfu_cache.put(req, self.current_sim_time)

            gds_res = self.gds_cache.get(req_id, self.current_sim_time)
            if not gds_res["hit"]:
                self.gds_cache.put(req, self.current_sim_time)

        # 3. Database Simulation
        db_telemetry = self.db_simulator.process_misses(tick_misses, delta_ms, workload_type)

        # 4. Backend Service Latency Generation
        backend_result = self.backend_service.process_request_batch(batch_results, db_telemetry, workload_type)
        for lat in backend_result["latencies"]:
            self.latency_reservoir.append(lat)

        # 5. Periodic Smart Cache Maintenance
        self.smart_cache.periodic_maintenance(self.current_sim_time, len(requests))

        # 6. Predictive Pre-Warming Evaluation
        traffic_status = self.traffic_monitor.record_tick(effective_rps, scenario)
        catalog_items = self.catalog.get_catalog(self.workload_generator.workload_type)
        resident_keys = set(self.smart_cache.entries.keys())
        
        prewarm_candidates = self.predictive_warmer.evaluate_prewarm_candidates(
            catalog_items,
            resident_keys,
            self.current_sim_time,
            traffic_status["trendVelocity"]
        )
        for cand in prewarm_candidates:
            self.smart_cache.put(cand["item"], self.current_sim_time)
            self.decision_logger.log_cache_action(
                action=DecisionType.PRE_WARM.value,
                item_id=cand["item"]["id"],
                score=cand["score"],
                reason=cand["reason"],
                factors={"trend": 0.95, "retrievalCost": 0.90},
                timestamp=self.current_sim_time
            )

        # 7. Compute Telemetry, Percentiles & Costs
        sorted_lats = sorted(list(self.latency_reservoir))
        n_lats = len(sorted_lats)
        p50 = sorted_lats[int(n_lats * 0.50)] if n_lats > 0 else 4.0
        p95 = sorted_lats[int(n_lats * 0.95)] if n_lats > 0 else 18.0
        p99 = sorted_lats[int(n_lats * 0.99)] if n_lats > 0 else 45.0
        avg_lat = (sum(sorted_lats) / float(n_lats)) if n_lats > 0 else 6.0

        hit_rate = self.smart_cache.get_hit_rate()
        miss_rate = 1.0 - hit_rate

        cost_telemetry = {
            "cacheCapacityBytes": self.cache_capacity_bytes,
            "usedBytes": self.smart_cache.used_bytes,
            "hitsPerSecond": effective_rps * hit_rate,
            "missesPerSecond": effective_rps * miss_rate,
            "backendLoadPercent": backend_result["backendLoadPercent"],
            "dbQueriesPerSecond": effective_rps * miss_rate
        }
        cost_result = self.cost_model.compute_cost(cost_telemetry)

        # 8. Adaptive Scaling Evaluation
        scaling_telemetry = {
            "hitsPerSecond": cost_telemetry["hitsPerSecond"],
            "missesPerSecond": cost_telemetry["missesPerSecond"],
            "hitRate": hit_rate,
            "backendLoadPercent": backend_result["backendLoadPercent"],
            "memoryUsagePercent": self.smart_cache.get_usage_percent()
        }
        scaling_result = self.scaling_controller.evaluate_scaling(
            self.cache_capacity_bytes,
            scaling_telemetry,
            scenario=scenario,
            current_time=self.current_sim_time
        )

        # Handle automated capacity scaling if approved and cooldown passed
        if (self.current_sim_time - self.last_scaling_action_time > 20.0):
            if scaling_result["shouldScaleUp"]:
                new_bytes = int(scaling_result["proposedGB"] * 1024 * 1024 * 1024)
                self.set_capacity(new_bytes)
                self.last_scaling_action_time = self.current_sim_time
                self.decision_logger.log_decision(
                    event_type=DecisionType.SCALE_UP.value,
                    title=f"Scale-Up Executed: {scaling_result['currentGB']} GB ➔ {scaling_result['proposedGB']} GB",
                    description=scaling_result["decisionReason"],
                    severity="primary",
                    timestamp=self.current_sim_time
                )
            elif scaling_result["shouldScaleDown"]:
                new_bytes = int(scaling_result["proposedGB"] * 1024 * 1024 * 1024)
                self.set_capacity(new_bytes)
                self.last_scaling_action_time = self.current_sim_time
                self.decision_logger.log_decision(
                    event_type=DecisionType.SCALE_DOWN.value,
                    title=f"Scale-Down Executed: {scaling_result['currentGB']} GB ➔ {scaling_result['proposedGB']} GB",
                    description=scaling_result["decisionReason"],
                    severity="info",
                    timestamp=self.current_sim_time
                )

        pollution_status = self.smart_cache.pollution_guard.get_status()

        snapshot = {
            "simTime": round(self.current_sim_time),
            "trafficRps": round(effective_rps, 1),
            "hitRatePercent": round(hit_rate * 100.0, 1),
            "missRatePercent": round(miss_rate * 100.0, 1),
            "p50LatencyMs": round(p50, 1),
            "p95LatencyMs": round(p95, 1),
            "p99LatencyMs": round(p99, 1),
            "avgLatencyMs": round(avg_lat, 1),
            "costPerHour": cost_result["totalCostPerHour"],
            "costSavingsPerHour": cost_result["costSavingsPerHour"],
            "savingsPercentage": cost_result["savingsPercentage"],
            "uncachedCostPerHour": cost_result["uncachedTotalCostPerHour"],
            "cacheCostPerHour": cost_result["cacheCostPerHour"],
            "computeCostPerHour": cost_result["computeCostPerHour"],
            "dbCostPerHour": cost_result["dbCostPerHour"],
            "backendLoadPercent": backend_result["backendLoadPercent"],
            "activeThreads": backend_result["activeThreads"],
            "dbCpuPercent": db_telemetry["cpuUtilizationPercent"],
            "dbLatencyMs": db_telemetry["currentLatencyMs"],
            "dbConnections": db_telemetry["activeConnections"],
            "dbQueriesPerSecond": db_telemetry["queriesPerSecond"],
            "memoryUsedMB": round(self.smart_cache.used_bytes / (1024.0 * 1024.0), 1),
            "memoryCapacityMB": round(self.cache_capacity_bytes / (1024.0 * 1024.0), 1),
            "memoryUsagePercent": round(self.smart_cache.get_usage_percent(), 1),
            "evictions": self.smart_cache.evictions,
            "refreshes": self.smart_cache.refreshes,
            "activeItemsCount": len(self.smart_cache.entries),
            "workloadType": workload_type,
            "scenario": scenario,
            "trafficState": traffic_status,
            "pollutionRisk": pollution_status["riskLevel"],
            "uniqueKeyRatePercent": pollution_status["uniqueKeyRatePercent"],
            "usefulOccupancyPercent": pollution_status["usefulOccupancyPercent"],
            "protectedItemsCount": pollution_status["protectedItemsCount"],
            "scalingDecision": scaling_result
        }

        self.rolling_history.append(snapshot)

        return {
            "snapshot": snapshot,
            "events": self.decision_logger.get_recent_events(30),
            "objects": [e.to_dict() for e in self.smart_cache.get_entries_list()[:60]],
            "battle": {
                "smart": len(self.smart_cache.entries),
                "lru": len(self.lru_cache.entries),
                "lfu": len(self.lfu_cache.entries),
                "gds": len(self.gds_cache.entries)
            }
        }
